### ISSUE TITLE
Following large files (docker logs -f) does not work as expected when combined with --tail or --since on the json-file logger

### ISSUE THREAD

**Issue by lukebayes** [docker/docker#21265]
If the `-f` flag is present, seeking begins at the start of the file, even if `--since` or `--tail` are provided.

I have a log file that has grown quite large.

If I run `docker logs --tail 100`, it returns right away with the expected lines, even if the log file is huge.

If I run `docker logs --since 2m`, it returns right away with the expected lines, even if the log file is huge.

If I run `docker logs -f`, I will never get to the end of the log file because it begins sending bits from the beginning of the file.

If I run `docker logs --tail 100 -f` or `docker logs --since 2m -f`, against a very large log file, the process hangs, cpu is maxed and eventually if it can get enough resources, it behaves as expected.

Unfortunately, the typical conditions where we need this feature to work (for our product at least), tends to be when we're having resource contention on a given node and we want to understand it.

This problem makes it impossible for us to get recent context and continue watching logs as we interact with the application.

**Output of `docker version`:**

```
Client:
 Version:      1.8.3
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   cedd534-dirty
 Built:        Thu Feb 18 16:25:16 UTC 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.8.3
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   cedd534-dirty
 Built:        Thu Feb 18 16:25:16 UTC 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 140
Images: 807
Storage Driver: overlay
 Backing Filesystem: extfs
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 4.2.2-coreos-r2
Operating System: CoreOS 835.13.0
CPUs: 8
Total Memory: 31.43 GiB
Name: [redacted]
ID: [redacted]
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

AWS EC2 and Kubernetes

**Steps to reproduce the issue:**
If I run `docker logs -f`, all lines are returned from the beginning of the file (over a very long time if the log file is very large). This seems reasonable, but in production, this command is almost always paired with `--tail` or `--since`, but whenever `-f` is present, the process clearly seeks through the entire log file before returning the requested lines.

If I run `docker logs --tail 100`, the last hundred lines are returned, but the request exits and the file is not tailed.

On a machine that is currently experiencing load problems and log spew (which is at least one compelling reason to use the `logs` feature), this command causes the process to hang indefinitely.

I suspect the problem is related to the order in which we work with inbound arguments in the [jsonfilelog/read.go](https://github.com/docker/docker/blob/a7fefcf16cda4f91b9526b4db6d06ef12652f419/daemon/logger/jsonfilelog/read.go#L88) file.

**Comment by cpuguy83** [comment]
How big is a large file?

**Comment by sravanikoilada** [comment]
it is working fine for me..

**Comment by AndrewGuenther** [comment]
If your file is that large shouldn't you be rotating it?

**Comment by lukebayes** [comment]
@cpuguy83  Large was ~5GB
@AndrewGuenther Yep, we worked around the problem (for now) by rotating logs more frequently

@sravanikoilada, et al: The bug is that `-f` breaks the expected behavior that `--tail` implies based on experiences with tools like `grep`.

The expectation is that these flags begin scanning a file of any size from the end, going backward, but in this case, whenever the `-f` flag is present, these expectations are not met and instead, the file is scanned from the beginning.

I believe this is a serious bug, because there are a number of problems that can cause a log file to grow  relatively large that don't necessarily implicate inappropriate configuration (though that was, admittedly the case here) and that these problems tend to coincide with environments that are resource constrained or otherwise under load.

This particular bug is especially insidious because it is most likely to surface on a system at exactly the point in time when one is attempting to understand problems related to load.

**Comment by cpuguy83** [comment]
@lukebayes `-f` does not cause docker to scan from the beginning. `-f` tells docker to follow the file, and you need to set `--tail` if you don't want to start from the beginning.

**Comment by cpuguy83** [comment]
Also note, we look at _all_ the files not just the current one.

**Comment by lukebayes** [comment]
I could be misunderstanding the root cause and jumping to incorrect conclusions, so please forgive me if that's the case.

The observed facts with at least one large log file are:
1)  `--tail` works perfectly alone
2) `--since` works perfectly alone
3) Introducing `-f` causes the process to lag and become unresponsive

**Comment by lukebayes** [comment]
Given the updated information, any additional thoughts on this?

**Comment by AndrewGuenther** [comment]
Ping @cpuguy83 
It seems like there is an actual issue here when combining the flags. My local tests see the same behavior.

**Comment by cpuguy83** [comment]
All I can say is try with master, some of this code was recently refactored.
The combination of the flags should not matter at all.
`-f` means we make one more function call at the end of the logs function. So `-f` alone should have the same problem as `-f` combined with other things.

**Comment by cpuguy83** [comment]
`-f` is itself a bit more efficient and reliable, in addition squashed some issues where we weren't closing out opened log files until following was stopped.

**Comment by cpuguy83** [comment]
Closing this stale issue.

**Comment by Gui13** [comment]
It is not stale, there is still the issue as of today.

**Comment by cpuguy83** [comment]
@Gui13 Can you open a new issue with a detailed description?
Thanks.
Please also make sure you are using an update to date version of Docker.

Thanks!

**Comment by csonuryilmaz** [comment]
@Gui13 I'm using Docker version 18.09.0 on CentOS 7. As @lukebayes mentioned if you use `-f` with `--tail` logs don't start from beginning. For example below command works for me:
```bash
docker logs -f --tail 10 gitlab
```

**Comment by Gui13** [comment]
Yep it seems that it is working, I don’t exactly know what I got 4 months ago but I can now follow with tail.

**Comment by estani** [comment]
In `Docker version 19.03.5, build 633a0ea838`

```
docker logs -f --tail 10 gitlab
```

just shows the last 10 entries and stops. `--since` seems to work fine though and honouring the `-f` flag.

This seem like a bug in that version at least, but I couldn't find any reference to it...
